### PR TITLE
[RISCV] fix SP recovery in varargs functions

### PR DESCRIPTION
**Pull request by daniilavdeev**
This patch fixes sp recovery in the epilogue in varargs functions when fp register is presented and second sp adjustment is applied.

Source of the issue: https://github.com/llvm/llvm-project/pull/110809

### CODE REVIEW

**Comment by dlav-sc** [comment]
@eaeltsin FYI

File: llvm/test/CodeGen/RISCV/pr114316.ll [hunk 1]
```diff
@@ -0,0 +1,53 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -riscv-v-vector-bits-max=512 -mtriple=riscv64 -mattr=+m,+c,+v -O0 < %s | FileCheck --check-prefix=RV64V %s
+
+; Function Attrs: nocallback nofree nosync nounwind willreturn
```
Comment (jrtc27) [inline llvm/test/CodeGen/RISCV/pr114316.ll:4]: Can you reduce the attributes (and remove the pointless Function Attrs comments)? I imagine you don't need most of them.

Comment (dlav-sc) [reply llvm/test/CodeGen/RISCV/pr114316.ll]: addressed

Comment (jrtc27) [reply llvm/test/CodeGen/RISCV/pr114316.ll]: Thanks, looks much better now (without actually digging into the details of the change)

**Comment by jrtc27** [comment]
Do you plan to precommit the test to show the diff for making the fix?

**Comment by dlav-sc** [comment]
> Do you plan to precommit the test to show the diff for making the fix? (Your commits here add the test _after_ the fix)

Yep, take a look, please

**Comment by eaeltsin** [comment]
Thank you for the fix!

**Comment by eaeltsin** [comment]
Can you please merge this soon? Thanks!

**Comment by lenary** [comment]
@eaeltsin there are no reviews yet, but you could provide one :)

I haven't (yet) because I am not super familiar with the code affected any more, but the commit and the change in the test look good enough to land, from my perspective.

**Comment by jrtc27** [comment]
Commit 1 should be a real commit to the tree, i.e. not squashed as part of merging this PR. Commits 2 and 3 should be one commit, i.e. squash and merge is right here. Though it's a bit odd you chose to split them out into separate commits in the PR in the first place.

**Comment by dlav-sc** [comment]
> Commit 1 should be a real commit to the tree, i.e. not squashed as part of merging this PR.

I've made a separate PR: https://github.com/llvm/llvm-project/pull/114970

**Comment by kito-cheng** [comment]
It's LGTM and I will give formal LGTM once https://github.com/llvm/llvm-project/pull/114970 land and this PR rebased

**Comment by lenary** [comment]
> > Commit 1 should be a real commit to the tree, i.e. not squashed as part of merging this PR.
> 
> I've made a separate PR with test: #114970

For future reference, this is something you can directly commit to `main` without a PR, like before we used github (it's a new test, and should be NFC and passing when committed).

**Review by kito-cheng — APPROVED** [review]
LGTM

**Comment by dlav-sc** [comment]
Thanks for taking a look at the patch!

**Comment by eaeltsin** [comment]
Thanks for the fix!

PS it is extremely unfortunate that this fix doesn't cherry-pick cleanly before https://github.com/llvm/llvm-project/pull/110810 - so we'll have to cherry-pick also the new functionality to get this fix :(

**Pull request merged**